EIPS/eip-3540.md

```diff
@@ -0,0 +1,218 @@
+
+At `block.number == HF2_BLOCK` new contract creation is modified:
+- if *initcode* or *code* starts with the *EOF prefix* (*TBD*), it is considered to be EOF formatted and will undergo validation specified in the following sections,
+- if *code* starts with `0xEF`, creation continues to result in an exceptional abort (the rule introduced in HF1),
+- otherwise code is considered *legacy code* and the following rules do not apply to it.
+
+Note that the *EOF prefix* here means the to be determined combination of `FORMAT` and *magic*.
+
+#### Container specification
+
+The format starts with the header:
+
+| description | length  | value | |
+|-------------|---------|-------|-|
+| format      | 1-byte  | 0xEF  | |
+| magic       | n-byte(s)  | TBD   | n >= 0 (zero in the best case) |
+| version     | 1-byte | 0x01  | means EOF1 |
+
+This is followed by one or more section headers:
+
+| description  | length  | | 
+|--------------|---------|-|
+| section_kind | 1-byte  | Encoded as a 8-bit unsigned number. |
+| section_size | 2-bytes | Encoded as a 16-bit unsigned big-endian number. |
+
+The section kinds are defined as follows:
+
+| section_kind | meaning    |
+|--------------|------------|
+| 0            | terminator |
+| 1            | code       |
+| 2            | data       |
+
+If the terminator is encountered, section size MUST NOT follow.
+
+The section contents follow after the header, in the order and size they are defined, without any padding bytes.
+
+To summarise, the bytecode has the following format:
+```
+format, magic, version, (section_kind, section_size)+, 0, <section contents>
+```
+
+#### Validation rules
+
+A bytestream starting with the *EOF prefix* declares itself conforming to the rules according to its version.
+
+1. The rules of `version=1` are specified below:
+- `section_size` MUST NOT be 0.
+- Exactly one code section MUST be present.
+- The code section MUST be the first section.
+- A single data section MAY follow the code section.
+- Stray bytes outside of sections MUST NOT be present. This includes trailing bytes after the last section.
+2. Any other version is invalid.
+
+(*Note:* Contract creation code SHOULD set the section size of the data section so that the constructor arguments fit it.)
+
+#### Changes to execution semantics
+
+For clarity, the *container* refers to the complete account code, while *code* refers to the contents of the code section only.
+
+1. Jumpdest analysis is only run on the *code*.
+2. Execution starts at the first byte of the *code*, and `PC` is set to this position within the container format (e.g. `PC=10` for a *container* with a code and data section).
+3. If `PC` goes outside of the code section bounds, execution  aborts with failure.
+4. `PC` returns the current position within the *container*.
+5. `JUMP`/`JUMPI` uses an absolute offset within the *container*.
```

Comment:
Why is PC and JUMP container absolute rather than code relative?  Given that neither can point outside of the code, it seems like this will introduce more complexity because now you have to know the offset of the code within the container in order to properly construct the JUMPs and during execution as well.

Reply:
This is indeed a debatable point. The argument for PC/JUMP/JUMPI being relative to container start is that this keeps it consistent with the other opcodes (`codecopy` etc.). The downside is that an assembler needs to determine the header size before filling in the final jump targets. Another argument for keeping it as it is here is that it might be a smaller change to interpreter implementations.

Reply:
The reason it is kept this way is to reduce the impact on client implementations. The only change really is what section to run jumpdest analysis on, what is the starting PC and what is the "end of code" (i.e. the first 3 rules). 

Arguably it is easier to correctly change 2-3 assemblers, than changing all the clients correctly.

I personally was in favour of changing it already to use the code section only, but means we need to introduce `DATACOPY` in this version, which is a lot of more added complexity for little benefit.